File: packages/py-parity-checking/src/run-tspice/domains/geometry-gf.ts
import type {
  StepGeometryGfGfdist,
  StepGeometryGfGfrefn,
  StepGeometryGfGfrepf,
  StepGeometryGfGfrepi,
  StepGeometryGfGfsep,
  StepGeometryGfGfsstp,
  StepGeometryGfGfstep,
  StepGeometryGfGfstol,
  StepOutput,
} from "../../case-types.js";
import { getOrCreateWindow, requireWindow, type RunTspiceContext } from "../context.js";

type GeometryGfStep =
  | StepGeometryGfGfdist
  | StepGeometryGfGfrefn
  | StepGeometryGfGfrepf
  | StepGeometryGfGfrepi
  | StepGeometryGfGfsep
  | StepGeometryGfGfsstp
  | StepGeometryGfGfstep
  | StepGeometryGfGfstol;

/** Execute one `geometry-gf.*` workflow step in tspice. */
export function runGeometryGfStep(context: RunTspiceContext, step: GeometryGfStep): StepOutput {
  switch (step.op) {
    case "geometry-gf.gfsstp":
      context.spice.raw.gfsstp(step.step);
      return { op: step.op, value: null };

    case "geometry-gf.gfstep":
      return { op: step.op, value: context.spice.raw.gfstep(step.time) };

    case "geometry-gf.gfstol":
      context.spice.raw.gfstol(step.value);
      return { op: step.op, value: null };

    case "geometry-gf.gfrefn":
      return { op: step.op, value: context.spice.raw.gfrefn(step.t1, step.t2, step.s1, step.s2) };

    case "geometry-gf.gfrepi": {
      const window = requireWindow(context, step.windowId);
      context.spice.raw.gfrepi(window, step.begmss, step.endmss);
      return { op: step.op, value: null };
    }

    case "geometry-gf.gfrepf":
      context.spice.raw.gfrepf();
      return { op: step.op, value: null };

    case "geometry-gf.gfsep": {
      const cnfine = requireWindow(context, step.cnfineWindowId);
      const result = getOrCreateWindow(context, step.resultWindowId, Math.max(step.nintvls, 8));
      context.spice.raw.gfsep(
        step.targ1,
        step.shape1,
        step.frame1,
        step.targ2,
        step.shape2,
        step.frame2,
        step.abcorr,
        step.obsrvr,
        step.relate,
        step.refval,
        step.adjust,
        step.step,
        step.nintvls,
        cnfine,
        result,
      );
      return { op: step.op, value: null };
    }

    case "geometry-gf.gfdist": {
      const cnfine = requireWindow(context, step.cnfineWindowId);
      const result = getOrCreateWindow(context, step.resultWindowId, Math.max(step.nintvls, 8));
      context.spice.raw.gfdist(
        step.target,
        step.abcorr,
        step.obsrvr,
        step.relate,
        step.refval,
        step.adjust,
        step.step,
        step.nintvls,
        cnfine,
        result,
      );
      return { op: step.op, value: null };
    }

    default: {
      const exhaustive: never = step;
      throw new Error(`Unhandled geometry-gf step: ${JSON.stringify(exhaustive)}`);
    }
  }
}

File: packages/py-parity-checking/sidecars/spiceypy/src/py_parity_sidecar/models/geometry_gf.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal


@dataclass(frozen=True)
class StepGeometryGfGfsstp:
    op: Literal["geometry-gf.gfsstp"]
    step: float


@dataclass(frozen=True)
class StepGeometryGfGfstep:
    op: Literal["geometry-gf.gfstep"]
    time: float


@dataclass(frozen=True)
class StepGeometryGfGfstol:
    op: Literal["geometry-gf.gfstol"]
    value: float


@dataclass(frozen=True)
class StepGeometryGfGfrefn:
    op: Literal["geometry-gf.gfrefn"]
    t1: float
    t2: float
    s1: bool
    s2: bool


@dataclass(frozen=True)
class StepGeometryGfGfrepi:
    op: Literal["geometry-gf.gfrepi"]
    windowId: str
    begmss: str
    endmss: str


@dataclass(frozen=True)
class StepGeometryGfGfrepf:
    op: Literal["geometry-gf.gfrepf"]


@dataclass(frozen=True)
class StepGeometryGfGfsep:
    op: Literal["geometry-gf.gfsep"]
    targ1: str
    shape1: str
    frame1: str
    targ2: str
    shape2: str
    frame2: str
    abcorr: str
    obsrvr: str
    relate: str
    refval: float
    adjust: float
    step: float
    nintvls: int
    cnfineWindowId: str
    resultWindowId: str


@dataclass(frozen=True)
class StepGeometryGfGfdist:
    op: Literal["geometry-gf.gfdist"]
    target: str
    abcorr: str
    obsrvr: str
    relate: str
    refval: float
    adjust: float
    step: float
    nintvls: int
    cnfineWindowId: str
    resultWindowId: str
